torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -0,0 +1,233 @@
+    }
+
+    if (!options_.defer_prefetch_) {
+      prefetch();
+    }
+  }
+
+  ~DataLoader() {
+    join();
+  }
+
+  void reset(bool prefetch = true) {
+    shuttle_.drain();
+    sampler_.reset();
+    sequence_number_ = 0;
+    if (prefetch) {
+      this->prefetch();
+    }
+  }
+
+  void prefetch(size_t requested_jobs) {
+    while (requested_jobs-- > 0) {
+      if (auto index_batch = get_index_batch()) {
+        push_job(std::move(*index_batch));
+      } else {
+        shuttle_.exhausted();
+      }
+    }
+  }
+
+  void prefetch() {
+    prefetch(*options_.max_jobs_);
+  }
+
+  at::optional<Batch> next() {
+    at::optional<Batch> batch;
+    if (options_.workers_ > 0) {
+      // sequencer_->next(...).map(|r| r.batch)
+      auto result = sequencer_->next(
+          [this] { return shuttle_.pop_result(options_.timeout_); });
+      if (result) {
+        batch = std::move(result->batch);
+      }
+      prefetch(1);
+    } else if (auto index_batch = get_index_batch()) {
+      batch = dataset_->batch(std::move(*index_batch));
+    }
+    return batch;
+  }
+
+  template <typename LoopFunction>
+  void loop(LoopFunction function) {
+    while (auto maybe_batch = next()) {
+      function(*maybe_batch);
+    }
+    reset();
+  }
+
+  void join() {
+    if (joined_) {
+      return;
+    }
+    shuttle_.drain();
+    // Send one 'quit' message per worker. Since a worker dies (exits its
+    // thread) after receiving this message, each `QuitWorker()` message will be
+    // read by exactly one worker.
+    for (size_t w = 0; w < options_.workers_; ++w) {
+      push_job(QuitWorker());
+    }
+    for (auto& worker : workers_) {
+      worker.join();
+    }
+    joined_ = true;
+  }
+
+  const DataLoaderOptions& options() const noexcept {
+    return options_;
+  }
+
+  size_t dataset_size() const noexcept {
+    return dataset_size_;
+  }
+
+ private:
+  struct Sequenced {
+    Sequenced() = default;
+    Sequenced(size_t sqn) : sequence_number(sqn) {}
+    size_t sequence_number;
+  };
+
+  struct QuitWorker {};
+  // Job = Enum(QuitWorker, IndexBatch)
+  struct Job : Sequenced {
+    Job() = default;
+    Job(QuitWorker q, size_t sqn) : Sequenced(sqn), quit(q) {}
+    Job(IndexBatch&& i, size_t sqn)
+        : Sequenced(sqn), index_batch(std::move(i)) {}
+    at::optional<QuitWorker> quit;
+    at::optional<IndexBatch> index_batch;
+  };
+
+  struct Result : Sequenced {
+    Result() = default;
+    Result(Batch&& b, size_t sqn) : Sequenced(sqn), batch(std::move(b)) {}
+    Batch batch;
+  };
+
+  void worker_thread(DatasetType dataset) {
+    while (true) {
+      auto job = shuttle_.pop_job();
+      if (job.quit) {
+        break;
+      }
+      auto batch = dataset.batch(std::move(*job.index_batch));
+      shuttle_.push_result({std::move(batch), job.sequence_number});
+    }
+  }
+
+  at::optional<IndexBatch> get_index_batch() {
+    auto indices = sampler_.next(options_.batch_size_);
```

Comment:
FWIW there's a concept of a `BatchSampler` in Python, which lets you avoid artificially creating batches in every single of your samplers, and really consider it to be a stream of indices (not a stream of chunks of indices).